pyramid/csrf.py

```diff
@@ -0,0 +1,283 @@
+    :class:`pyramid.interfaces.ICSRFStoragePolicy` implementation against the value in
+    ``request.POST.get(token)`` (if a POST request) or
+    ``request.headers.get(header)``. If a ``token`` keyword is not supplied to
+    this function, the string ``csrf_token`` will be used to look up the token
+    in ``request.POST``. If a ``header`` keyword is not supplied to this
+    function, the string ``X-CSRF-Token`` will be used to look up the token in
+    ``request.headers``.
+
+    If the value supplied by post or by header doesn't match the value supplied
+    by ``policy.get_csrf_token()`` (where ``policy`` is an implementation of
+    :class:`pyramid.interfaces.ICSRFStoragePolicy`), and ``raises`` is ``True``, this
+    function will raise an :exc:`pyramid.exceptions.BadCSRFToken` exception. If
+    the values differ and ``raises`` is ``False``, this function will return
+    ``False``.  If the CSRF check is successful, this function will return
+    ``True`` unconditionally.
+
+    See :ref:`auto_csrf_checking` for information about how to secure your
+    application automatically against CSRF attacks.
+
+    .. versionadded:: 1.4a2
+
+    .. versionchanged:: 1.7a1
+       A CSRF token passed in the query string of the request is no longer
+       considered valid. It must be passed in either the request body or
+       a header.
+
+    .. versionchanged:: 1.8a1
+       Moved from pyramid.session to pyramid.csrf
+    """
+    supplied_token = ""
+    # We first check the headers for a csrf token, as that is significantly
+    # cheaper than checking the POST body
+    if header is not None:
+        supplied_token = request.headers.get(header, "")
+
+    # If this is a POST/PUT/etc request, then we'll check the body to see if it
+    # has a token. We explicitly use request.POST here because CSRF tokens
+    # should never appear in an URL as doing so is a security issue. We also
+    # explicitly check for request.POST here as we do not support sending form
+    # encoded data over anything but a request.POST.
+    if supplied_token == "" and token is not None:
+        supplied_token = request.POST.get(token, "")
+
+    policy = request.registry.queryUtility(ICSRFStoragePolicy)
+    if policy is None:
+        # There is no policy set, but we are trying to validate a CSRF token
+        # This means explicit validation has been asked for without configuring
+        # the CSRF implementation. Fall back to SessionCSRFStoragePolicy as that is the
+        # default
+        policy = SessionCSRFStoragePolicy()
+    if not policy.check_csrf_token(request, supplied_token):
+        if raises:
+            raise BadCSRFToken('check_csrf_token(): Invalid token')
+        return False
+    return True
+
+
+def check_csrf_origin(request, trusted_origins=None, raises=True):
+    """
+    Check the Origin of the request to see if it is a cross site request or
+    not.
+
+    If the value supplied by the Origin or Referer header isn't one of the
```

Comment:
ibid. So perhaps ``Origin`` and ``Referer`` (also ``Referrer``). Although the former misspelling seems to be inherited, I think the correct spelling is valid as well.